File: clients/desktop/src/main/task/schedule-manager.ts
/**
 * 调度管理器
 * 
 * 对等 Android ScheduleManager
 */

import * as fs from 'fs';
import * as path from 'path';
import { app } from 'electron';

export class ScheduleManager {
  private file: string;
  private schedules: Map<string, any> = new Map();

  constructor() {
    const dataDir = path.join(app.getPath('userData'), 'data');
    if (!fs.existsSync(dataDir)) {
      fs.mkdirSync(dataDir, { recursive: true });
    }
    this.file = path.join(dataDir, 'schedules.json');
    this.load();
  }

  private load() {
    try {
      if (fs.existsSync(this.file)) {
        const content = fs.readFileSync(this.file, 'utf-8');
        const obj = JSON.parse(content);
        for (const key in obj) {
          this.schedules.set(key, obj[key]);
        }
        console.log(`[ScheduleManager] Loaded ${this.schedules.size} schedules`);
      }
    } catch (error) {
      console.error('[ScheduleManager] Failed to load:', error);
    }
  }

  private save() {
    try {
      const obj: any = {};
      this.schedules.forEach((value, key) => {
        obj[key] = value;
      });
      fs.writeFileSync(this.file, JSON.stringify(obj, null, 2));
    } catch (error) {
      console.error('[ScheduleManager] Failed to save:', error);
    }
  }

  add(config: any): string {
    return this.addSync(config);
  }

  addSync(config: any): string {
    const id = config.id || `schedule_${Date.now()}`;
    config.id = id;
    this.schedules.set(id, config);
    this.save();
    console.log(`[ScheduleManager] Added schedule: ${id}`);
    return id;
  }

  update(id: string, config: any): void {
    this.updateSync(id, config);
  }

  updateSync(id: string, config: any): void {
    config.id = id;
    this.schedules.set(id, config);
    this.save();
    console.log(`[ScheduleManager] Updated schedule: ${id}`);
  }

  delete(id: string): void {
    this.deleteSync(id);
  }

  deleteSync(id: string): void {
    this.schedules.delete(id);
    this.save();
    console.log(`[ScheduleManager] Deleted schedule: ${id}`);
  }

  get(id: string): any | null {
    return this.schedules.get(id) || null;
  }

  list(): any[] {
    return Array.from(this.schedules.values());
  }
}

File: clients/app/android/app/src/main/java/com/weiqi/app/task/ScheduleManager.kt
package com.weiqi.app.task

import android.content.Context
import com.weiqi.app.util.Logger
import org.json.JSONObject
import java.io.File

/**
 * ScheduleManager - 定时计划管理器（单例）
 *
 * 职责：
 * - JSON 文件存储
 * - 增删改查接口
 * - 线程安全：所有读写操作通过 synchronized(schedules) 保护
 */
class ScheduleManager private constructor(context: Context) {
    
    companion object {
        private const val TAG = "ScheduleManager"
        private const val FILE_NAME = "schedules.json"
        
        @Volatile
        private var instance: ScheduleManager? = null
        
        fun getInstance(context: Context): ScheduleManager {
            return instance ?: synchronized(this) {
                instance ?: ScheduleManager(context.applicationContext).also { instance = it }
            }
        }
    }
    
    private val file = File(context.filesDir, FILE_NAME)
    private val schedules = JSONObject()
    
    init {
        load()
    }
    
    /**
     * 加载计划文件（构造时调用，无需锁）
     */
    private fun load() {
        try {
            if (file.exists()) {
                val content = file.readText()
                val loaded = JSONObject(content)
                val keys = loaded.keys()
                while (keys.hasNext()) {
                    val key = keys.next()
                    schedules.put(key, loaded.get(key))
                }
                Logger.i(TAG, "Loaded ${schedules.length()} schedules")
            } else {
                save()
            }
        } catch (e: Exception) {
            Logger.e(TAG, "Failed to load schedules", e)
        }
    }
    
    /**
     * 保存计划文件（调用方需持有 schedules 锁）
     */
    private fun save() {
        try {
            file.writeText(schedules.toString(2))
            Logger.d(TAG, "Saved schedules to file")
        } catch (e: Exception) {
            Logger.e(TAG, "Failed to save schedules", e)
        }
    }
    
    /**
     * 添加计划
     */
    fun add(config: JSONObject): String {
        val id = config.optString("id", "schedule_${System.currentTimeMillis()}")
        config.put("id", id)
        synchronized(schedules) {
            schedules.put(id, config)
            save()
        }
        Logger.i(TAG, "Added schedule: $id")
        return id
    }
    
    /**
     * 更新计划
     */
    fun update(id: String, config: JSONObject) {
        config.put("id", id)
        synchronized(schedules) {
            schedules.put(id, config)
            save()
        }
        Logger.i(TAG, "Updated schedule: $id")
    }
    
    /**
     * 删除计划
     */
    fun delete(id: String) {
        synchronized(schedules) {
            schedules.remove(id)
            save()
        }
        Logger.i(TAG, "Deleted schedule: $id")
    }
    
    /**
     * 获取计划
     */
    fun get(id: String): JSONObject? = synchronized(schedules) {
        schedules.optJSONObject(id)
    }
    
    /**
     * 列出所有计划
     */
    fun list(): List<JSONObject> = synchronized(schedules) {
        val list = mutableListOf<JSONObject>()
        val keys = schedules.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            schedules.optJSONObject(key)?.let { list.add(it) }
        }
        list
    }
}
